File: CS/Example_FormLayout/Controllers/HomeController.cs
using Example_FormLayout.Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace Example_FormLayout.Controllers
{
    public class HomeController : Controller
    {

        public ActionResult Index()
        {
            var model = new Category() { ID = 1, Name = "Fruits", Products = new List<Product>() };
            Session["CategoryData"] = model;
            return View(model);
        }

        [HttpPost]
        public ActionResult Index(Category cat) {
            return View(cat);
        }
        public ActionResult CategoryFormLayoutPartial(string command)
        {
            Category p = Session["CategoryData"] as Category;
            string[] results = command.Split(';');            
            string action = results[0];
            int index = Convert.ToInt32(results[1]);
            if (action == "add")
            {
                p.Products.Add(new Product() { ProductName = String.Empty });
            }
            else
            {
                p.Products.RemoveAt(index);
            }
            return PartialView(p);
        }
    }
}

File: VB/Example_FormLayout_VB/Controllers/HomeController.vb
Imports System.Web.Mvc

Public Class HomeController
    Inherits Controller

    Function Index() As ActionResult
        Dim model As Category = New Category() With {.ID = 1, .Name = "Fruits", .Products = New List(Of String)()}
        Session("CategoryData") = model
        Return View(model)
    End Function

    Function CategoryFormLayoutPartial(command As String) As ActionResult
        Dim p As Category = TryCast(Session("CategoryData"), Category)
        Dim results As String() = command.Split(";"c)
        Dim action As String = results(0)
        Dim index As Integer = Convert.ToInt32(results(1))
        If action = "add" Then
            p.Products.Add([String].Empty)
        Else
            p.Products.RemoveAt(index)
        End If
        Return PartialView(p)
    End Function


End Class


